src/Repository.ts

```diff
@@ -322,30 +331,46 @@ export class MongoRepository<DOC, DTO = DOC> {
     if (opts?.noClone === false || (opts?.noClone === undefined && this.options.eventOpts?.noClone !== true)) {
       // Dereference (aka clone) the target document(s) to
       // prevent down stream modifications in events on the original instance(s)
-      newDocument = clone(newDocument);
-      if (originalDocument) {
-        originalDocument = clone(originalDocument);
-      }
+      this.cloneDocuments(newDocument, originalDocument);
     }
 
     for (const fn of fns) {
       const events = Reflect.getMetadata(`${type}_${fn}`, this) || [];
-      for (const event of events) {
-        const repoEventArgs: RepoEventArgs = {
-          originalDocument,
-          operation: fn,
-          operationType: type
-        };
-        newDocument = event.bind(this)(newDocument, repoEventArgs);
-        if (typeof newDocument.then === 'function') {
-          newDocument = await newDocument;
-        }
-      }
+      await this.bindEvents(_.cloneDeep(events), fn, type, originalDocument, newDocument);
     }
-
     return newDocument;
   }
 
+  protected cloneDocuments(newDocument: any, originalDocument: any): void {
+    newDocument = _.cloneDeep(newDocument);
+    if (originalDocument) {
+      originalDocument = _.cloneDeep(originalDocument);
+    }
+  }
+
+  protected async bindEvents(
```

Comment:
I using this cloning because If I don’t used It some test cases will fail. and I changed this for to a recursive because I'm trying to avoid using a foor loop inside of other, and the recursion is more efficient.

Reply:
Array.pop and recursion is going to be slower than a for.loop in general. Because you are modifying the argument array instance (`events.pop()`), this is modifying the same instance up the call stack, creating unintended side-effects. Most likely the cause for tests to the fail. Using a for.loop would not require the array manipulation, nor cloning.

I personally don't see an issue with a second deep for.loop but you can keep it as a separate method if you like, I however would suggest this is a `private` method at this point.